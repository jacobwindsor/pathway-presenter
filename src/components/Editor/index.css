.editor-wrapper {
    position: relative;
    height: 100%;
    width: 100%;
    overflow: hidden;
}

.editor-wrapper .right-section {
    width: 70%;
    margin-left: calc(30% + 2rem);
    height: 100%;
}

.editor-wrapper .right-section .diagram-comp-wrapper {
    height: 80%;
    width: 100%;
}

.editor-wrapper .add-slide-button {
    position: absolute;
    bottom: 1.6rem;
    right: 2rem;
}