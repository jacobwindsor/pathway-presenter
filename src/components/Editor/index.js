import './index.css';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import presentations from '../../data/presentations';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import Adder from './components/Adder';
import Creator from './components/Creator';
import ErrorMessage from '../ErrorMessage';

class Editor extends Component {
    constructor(props) {
        super(props);
        this.state = {
            presentation: null,
            loading: true,
            error: null
        };
    }

    getPresentation = id => {
        presentations.get(id)
            .then(presentation => {
                this.setState({
                    presentation,
                    loading: false,
                })
            })
            .catch(err => {
                this.setState({
                    error: err,
                    loading: false,
                })
            })
    };

    componentDidMount() {
        const { presId } = this.props;
        if (! presId) return;
        this.getPresentation(presId);
    }

    savePresentation = presentation => {
        let promise;
        if(presentation.id) {
            promise = presentations.update(presentation.id, presentation)
        }
        else {
            promise = presentations.create(presentation);
        }

        promise
            .then(presentation => this.setState({presentation}))
            .catch(error => {
                this.setState({
                    error
                })
            })
    };

    handlePresentationCreate = toCreate => {
        this.setState({
            presentation: {
                wpId: toCreate.wpId,
                version: toCreate.version || 0,
                title: toCreate.title || null,
                description: toCreate.description || null,
                slides: [{
                    targets: [],
                }],
            }
        })
    };

    render() {
        const { loading, error, presentation } = this.state;

        const SuccessContent = () => {
            if (error) return null;
            return (
                <div className="full editor-wrapper">
                    { presentation ?
                        <Creator presentation={presentation} handleSave={this.savePresentation} /> :
                        <Adder handleSelect={this.getPresentation} handleCreate={this.handlePresentationCreate}/>
                    }
                </div>
            )
        };

        const ErrorContent = () => {
            if (! error) return null;
            return <ErrorMessage message={error.message} />
        };

        return (
            <MuiThemeProvider>
                <div className="full">
                    <ErrorContent />
                    <SuccessContent />
                </div>
            </MuiThemeProvider>
        )
    }
}

Editor.propTypes = {
    presId: PropTypes.string,
};

export default Editor;