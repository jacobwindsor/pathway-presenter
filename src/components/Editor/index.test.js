import React from 'react';
import { shallow } from 'enzyme';
import Editor from './index';
import toJson from 'enzyme-to-json';
import injectTapEventPlugin from 'react-tap-event-plugin';

describe('The Editor. This allows creation and updating of presentations', () => {
  beforeAll(() => {
    // Needed for onTouchTap
    // http://stackoverflow.com/a/34015469/988941
    injectTapEventPlugin();
  });

  it('Should render without failing', () => {
    const comp = shallow(<Editor presId="my-pres" />);
    expect(toJson(comp)).toMatchSnapshot();
  });

  it('should show a loading indicator', () => {
    const comp = shallow(<Editor presId="my-pres" />);
    expect(comp.dive().find('.loading-spinner')).toHaveLength(1);
  });

  it('should show the adder form', () => {
    const comp = shallow(<Editor />);
    comp.setState({
      loading: false,
      presentation: null
    });
    expect(comp.find('SuccessContent').dive().find('Adder')).toHaveLength(1);
    expect(
      toJson(comp.find('SuccessContent').dive().find('Adder'))
    ).toMatchSnapshot();
  });

  it('should show the creator interface', () => {
    const comp = shallow(<Editor presId="my-pres" />);
    comp.setState({
      loading: false,
      presentation: {
        title: 'My pres',
        authorName: 'Donald Duck',
        wpId: 'WP4',
        version: 0,
        slides: []
      }
    });
    expect(comp.find('SuccessContent').dive().find('Creator')).toHaveLength(1);
    expect(
      toJson(comp.find('SuccessContent').dive().find('Creator'))
    ).toMatchSnapshot();
  });

  it('should show an error', () => {
    const comp = shallow(<Editor presId="my-pres" />);
    comp.setState({
      error: {
        message: 'An error'
      },
      loading: false
    });
    expect(comp.find('ErrorContent')).toHaveLength(1);
    expect(toJson(comp.find('ErrorContent'))).toMatchSnapshot();
    expect(
      comp.find('ErrorContent').dive().find('ErrorMessage').prop('message')
    ).toEqual('An error');
  });
});
