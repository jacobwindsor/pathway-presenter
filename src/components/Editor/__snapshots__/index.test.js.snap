// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Editor. This allows creation and updating of presentations Should render without failing 1`] = `<Component />`;

exports[`The Editor. This allows creation and updating of presentations should show an error 1`] = `<ErrorContent />`;

exports[`The Editor. This allows creation and updating of presentations should show the adder form 1`] = `
<Adder
  handleCreate={[Function]}
  handleSelect={[Function]}
/>
`;

exports[`The Editor. This allows creation and updating of presentations should show the creator interface 1`] = `
<Creator
  handleDelete={[Function]}
  handleSave={[Function]}
  presentation={
    Object {
      "authorName": "Donald Duck",
      "slides": Array [],
      "title": "My pres",
      "version": 0,
      "wpId": "WP4",
    }
  }
/>
`;
